### PR TITLE
[#164542759] Adds detection of jailbreaked or rooted device

### PR DESCRIPTION
**Pull request by Edigio**
Stop running app when the device is Jailbreaked (iOs) or rooted (Android)

### CODE REVIEW

**Comment by Edigio** [comment]
IMG_1FD61EDED89B-1
device-2019-04-16-142802

On iOs the popup-message is shown after the spashscreen, beacause the splashscreen was superimposed on the message.

File: package.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
     "bundle:android-release": "node node_modules/react-native/local-cli/cli.js bundle --platform android --dev false --entry-file index.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res/",
     "generate:locales": "ts-node --skip-project -O '{\"lib\":[\"es2015\"]}' scripts/make-locales.ts",
     "generate:content-definitions": "rimraf definitions/content && mkdir -p definitions/content && gen-api-models --api-spec https://raw.githubusercontent.com/teamdigitale/italia-services-metadata/master/definitions.yml --out-dir ./definitions/content",
-    "generate:all": "npm-run-all generate:api-definitions generate:pagopa-api-definitions generate:content-definitions generate:locales"
+    "generate:all": "npm-run-all generate:api-definitions generate:pagopa-api-definitions generate:content-definitions generate:locales",
+    "build:ios": "react-native bundle --entry-file='index.js' --bundle-output='./ios/main.jsbundle' --dev=false --platform='ios'"
```
Comment (cloudify) [inline package.json]: @Edigio is this needed for the fix?

File: tsconfig.json [hunk 2]
```diff
@@ -1,5 +1,7 @@
 {
   "compilerOptions": {
+    "baseUrl": ".",
```
Comment (cloudify) [inline tsconfig.json]: changes to `tsconfig.json` shouldn't be necessary, are they?

File: package.json (hunk 1 above)
Comment (Edigio) [reply package.json]: :O no, it is not necessary sorry

File: ts/screens/IngressScreen.tsx [hunk 3]
```diff
@@ -43,7 +47,24 @@ class IngressScreen extends React.PureComponent<Props> {
   public componentDidMount() {
     // Dispatch START_APPLICATION_INITIALIZATION to initialize the app
     this.props.dispatch(startApplicationInitialization());
```
Comment (fpersico) [inline ts/screens/IngressScreen.tsx]: @cloudify @Edigio are we sure we want to start the applicationInizialization also if the device is rooted? I think it must be done only if not rooted.

Comment (cloudify) [reply ts/screens/IngressScreen.tsx]: agreed

**Comment by fpersico** [comment]
@Edigio is quite strange that a native library is not requiring any change on android and ios projects file. Are you sure all we need is already committed?

File: ts/screens/IngressScreen.tsx (hunk 3 above)
Comment (Edigio) [reply ts/screens/IngressScreen.tsx]: Ok, this is the result:
IMG_0225

**Comment by cloudify** [comment]
@Edigio there's a few changes not related to this PR (e.g. `android-sourcemap.*`, the font files, `Info.plist`, ...) pls don't include those changes - see [number 4](https://github.com/teamdigitale/io-handbook/blob/master/development-guidelines.md#general-guidelines)

File: ts/screens/IngressScreen.tsx [hunk 4]
```diff
@@ -41,9 +45,27 @@ const styles = StyleSheet.create({
  */
 class IngressScreen extends React.PureComponent<Props> {
   public componentDidMount() {
-    // Dispatch START_APPLICATION_INITIALIZATION to initialize the app
-    this.props.dispatch(startApplicationInitialization());
-
+    // if device is rooted and it isn't an emulator
+    if (JailMonkey.isJailBroken() && !DeviceInfo.isEmulator()) {
```
Comment (Undermaken) [inline ts/screens/IngressScreen.tsx]:
```suggestion
    if (!DeviceInfo.isEmulator() && JailMonkey.isJailBroken()) {
```

File: ts/screens/IngressScreen.tsx [hunk 5]
```diff
@@ -41,9 +45,27 @@ const styles = StyleSheet.create({
...
+      Alert.alert(
+        I18n.t("startup.rootedDeviceAlert.msgTitle"),
+        I18n.t("startup.rootedDeviceAlert.msgText"),
+        [
+          {
+            text: I18n.t("global.buttons.ok"),
+            style: "destructive",
+            onPress: () => {
+              // after confirmation close app
+              RNExitApp.exitApp();
```
Comment (Undermaken) [inline ts/screens/IngressScreen.tsx]: Is a library needed to exit the application?
What do you think @cloudify? [https://facebook.github.io/react-native/docs/backhandler#exitapp](url)

**Comment by cloudify** [comment]
@Edigio pls fix the build

File: ts/screens/IngressScreen.tsx (hunk 5 above)
Comment (fpersico) [reply ts/screens/IngressScreen.tsx]: @Undermaken BackHandler only works for for Android (end partially for TVOS).

**Comment by fpersico** [comment]
@Edigio add the missing updated yarn.lock.

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#164542759](https://www.pivotaltracker.com/story/show/164542759): APP_14/should: Jailbreak and root detection


New dependencies added: <a href='https://www.npmjs.com/package/jail-monkey'><code>jail-monkey</code></a> and <a href='https://www.npmjs.com/package/react-native-exit-app'><code>react-native-exit-app</code></a>.


<h2><a href="https://github.com/GantMan/jail-monkey#readme">jail-monkey</a></h2>
<p>Author: Gant Laborde</p>
<p>Description: A React Native module for identifying jail-broken, rooted, or mock locations on iOS and Android</p>
<p>Homepage: <a href="https://github.com/GantMan/jail-monkey#readme">https://github.com/GantMan/jail-monkey#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>over 2 years ago</td></tr><tr><td>Last Updated</td><td>14 days ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>14</td></tr><tr><td>Keywords</td><td>React-Native, Native, Jail, Break, Jail-Break, iOS, Android, Mock-Locations and Detect</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# WARNING:  I don't have the devices to test anymore, so testing is done by those submitting PRs bona fide.

Jail Monkey
## Can you ever really trust a phone?

# Why?
Are users claiming they are crossing the globe in seconds and collecting all the Pokeballs?  Some apps need to protect themselves in order to protect data integrity.  JailMonkey allows you to: 
* Identify if a phone has been jail-broken or rooted for iOS/Android.  
* Detect mocked locations for phones set in "developer mode".
* **(ANDROID ONLY)** Detect if the application is running on external storage such as an SD card.

# Use
```javascript
import JailMonkey from 'jail-monkey'

// is this device JailBroken on iOS/Android?
JailMonkey.isJailBroken()

// Can this device mock location - no need to root!
JailMonkey.canMockLocation()

// Check if device violates any of the above
JailMonkey.trustFall()

// (ANDROID ONLY) Check if application is running on external storage
JailMonkey.isOnExternalStorage()

// (ANDROID ONLY) Check if the phone has some malicious apps installed
JailMonkey.hookDetected()

// Check if the application is running in debug mode
JailMonkey.isDebugged()
```
Circle of Trust

### :exclamation: Since emulators are usually rooted, you might want to bypass these checks during development.  Unless you're keen on constant false alarms :alarm_clock:

# Install

```
npm i jail-monkey --save
react-native link
```

If you use `rnpm`, you may have trouble as `rnpm` does not link Android properly after 0.29.0!

# Additional Info
This has been made public to help keep it up to date.  As detection measures get better or out-dated, please send updates to this project so it can be the best method of detection.

Special thanks to this fantastic blog article:  http://blog.geomoby.com/2015/01/25/how-to-avoid-getting-your-location-based-app-spoofed/


</details>




<h2><a href="http://watchwimswork.be">react-native-exit-app</a></h2>
<p>Author: Wumke</p>
<p>Description: Exit,close,kill,shutdown app completely for React Native on iOS and Android.</p>
<p>Homepage: <a href="http://watchwimswork.be">http://watchwimswork.be</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>almost 2 years ago</td></tr><tr><td>Last Updated</td><td>over 1 year ago</td></tr><tr><td>License</td><td>GPLv3</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>2</td></tr><tr><td>Keywords</td><td>react, native, exit, close, shutdown, kill, android and ios</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

# react-native-exit-app
Exit / Close / Kill / shutdown your react native app. Does not invoke a crash notification.

NOTICE:
- for React Native < 0.47 use react-native-immediate-phone-call <1.x.x
- for React Native > 0.47 use react-native-immediate-phone-call >=1.x.x

## Setup

Fast and easy:
```bash
npm install react-native-exit-app --save
react-native link react-native-exit-app
```

Or manual: add the latest version as dependeny to your package.json.

```javascript
{
  "name": "YourProject",
  ...
  },
  "dependencies": {
    ...
    "react-native-exit-app": "0.1.0",
    ...
  }
```

#### iOS
* Add RNExitApp.xcoderproj into your project in the Libraries folder.
* Add the .a file on the General tab of your target under Linked Frameworks And Libraries
* Add the .a file on the Build Phases tab of your target under Link Binary With Libraries

#### Android
* In the settings.gradle
  ```
    include ':react-native-exit-app', ':app'
    project(':react-native-exit-app').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-exit-app/android')
  ```
* In the build.gradle
  ```
    compile project(':react-native-exit-app')
  ```
* In MainApplication.java
  ```
    import com.github.wumke.RNExitApp.RNExitAppPackage;
    ...
    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
        ...
        new RNExitAppPackage(),
        ...
      );
    }
    ...
  ```
## Usage

```javascript
import RNExitApp from 'react-native-exit-app';
...
RNExitApp.exitApp();
...
```

## Versioning

This project uses semantic versioning: MAJOR.MINOR.PATCH.
This means that releases within the same MAJOR version are always backwards compatible. For more info see [semver.org](http://semver.org/).


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,14755 @@
+{
```
Comment (cloudify) [inline package-lock.json]: don't add `package-lock.json`, we have `yarn.lock` instead

**Comment by Edigio** [comment]
device-2019-05-03-124121
IMG_0226
On ios i don't add the 'ok-close' button beacuse the exit from the app is seen as crash

File: android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java [hunk 7]
```diff
@@ -25,28 +29,84 @@ protected String getMainComponentName() {
...
+                }
+            };
+        }
     }
 
     @Override
     public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
-      CalendarEventsPackage.onRequestPermissionsResult(requestCode, permissions, grantResults);
-      super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        CalendarEventsPackage.onRequestPermissionsResult(requestCode, permissions, grantResults);
```
Comment (cloudify) [inline android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java]: looks like this file had 2 spaces indentation, keep that

File: android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java [hunk 8]
```diff
@@ -25,28 +29,84 @@ protected String getMainComponentName() {
...
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+    }
+
+    /**
+     * Check rooted device
+     */
+    private boolean isDeviceRooted() {
+        //check only once
+        if (isRootedDeviceFlag == null) {
+            isRootedDeviceFlag = (checkRootMethod1() || checkRootMethod2() || checkRootMethod3());
```
Comment (cloudify) [inline android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java]: pls add a comment to some reference documentation that explains these detection methods

File: android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java [hunk 9]
```diff
@@ -67,27 +67,30 @@ protected ReactRootView createRootView() {
...
     }
 
     private boolean checkRootMethod1() {
+        //check 1: get from build info, test-keys means it was signed with a custom key generated by a third-party developer
         String buildTags = android.os.Build.TAGS;
         return buildTags != null && buildTags.contains("test-keys");
     }
 
     private boolean checkRootMethod2() {
+        // check 2: if /system/app/Superuser.apk or its directories are presents
```
Comment (cloudify) [inline android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java]:
```
if /system/app/Superuser.apk or its directories are present
```

File: android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java [hunk 10]
```diff
@@ -99,6 +102,7 @@ private boolean checkRootMethod2() {
     private boolean checkRootMethod3() {
         Process process = null;
         try {
+            // check 3: try executing command, this operation require root
```
Comment (cloudify) [inline android/app/src/main/java/it/teamdigitale/app/italiaapp/MainActivity.java]:
```
try executing a command, this operation requires root privileges
```

**Comment by cloudify** [comment]
@fpersico @Undermaken please review

**Comment by fpersico** [comment]
@cloudify @cloudify Please also consider using already tested dection libraries:

iOS
https://github.com/thii/DTTJailbreakDetection

Android
https://github.com/scottyab/rootbeer

**Comment by Edigio** [comment]
> @cloudify @cloudify Please also consider using already tested dection libraries:
> 
> iOS
> https://github.com/thii/DTTJailbreakDetection
> 
> Android
> https://github.com/scottyab/rootbeer

@fpersico 
The ios library should perform the same check proposed in this PR, the android library instead takes its cue from the same solution found on stackoverflow and present in this PR.

Schermata 2019-05-06 alle 16 28 28

It seemed too much to add two libraries in the project for this story :)

**Comment by Edigio** [comment]
@fpersico @cloudify For android i added another control when the app running in the android emulator or genymotion emulator. Without this control we can't tested the app with genymotion emulator

**Pull request merged**